File: client/src/lib/ws/messages.ts
import type Card from "$lib/components/Card.svelte";
import type { GameVersionKey } from "$lib/engine/types";

export const OUT_MESSAGES = {
  START_GAME: 'START_GAME',
  CREATE_ROOM: 'CREATE_ROOM',
  JOIN_ROOM: 'JOIN_ROOM',
  CHECK_SET: 'CHECK_SET',
} as const;

export interface StartGameMessage {
  readonly type: typeof OUT_MESSAGES.START_GAME;
  roomID: string;
  gameVersion: GameVersionKey;
}

export interface CreateRoomMessage {
  readonly type: typeof OUT_MESSAGES.CREATE_ROOM;
  nickname: string;
}

export interface JoinRoomMessage {
  readonly type: typeof OUT_MESSAGES.JOIN_ROOM;
  roomID: string;
  nickname: string;
}

export interface CheckSetMessage {
  readonly type: 'CHECK_SET';
  cardIDs: string[];
  playerID: string;
  roomID: string;
  gameID: string;
}

export type OutMessage =
  | StartGameMessage
  | CreateRoomMessage
  | JoinRoomMessage
  | CheckSetMessage;





export const IN_MESSAGES = {
  CREATED_ROOM: 'CREATED_ROOM',
  JOINED_ROOM: 'JOINED_ROOM',
  LEFT_ROOM: 'LEFT_ROOM',
  RECONNECTED_TO_ROOM: 'RECONNECTED_TO_ROOM',
  SEND_STATE_TO_RECONNECTED: 'SEND_STATE_TO_RECONNECTED',
  STARTED_GAME: 'STARTED_GAME',
  CHECK_SET_RESULT: 'CHECK_SET_RESULT',
  CHANGED_GAME_STATE: 'CHANGED_GAME_STATE',
  GAME_OVER: 'GAME_OVER',
  ERROR: 'ERROR'
} as const;

export type RoomMember = { id: string; nickname: string; isConnected: boolean }
export type Player = RoomMember & { score: number }

export interface StartedGameMessage {
  readonly type: typeof IN_MESSAGES.STARTED_GAME;
  gameID: string;
  gameVersion: GameVersionKey;
  deck: Card[]
  players: Record<string, Player>;
}

export interface CreatedRoomMessage {
  readonly type: typeof IN_MESSAGES.CREATED_ROOM;
  roomID: string;
  playerID: string;
  nickname: string;
}

export interface JoinedRoomMessage {
  readonly type: typeof IN_MESSAGES.JOINED_ROOM;
  roomID: string;
  playerID: string;
  nickname: string;
  players: Player[];
}

export interface LeftRoomMessage {
  readonly type: typeof IN_MESSAGES.LEFT_ROOM;
  playerID: string;
}

export interface SendStateToReconnectedMessage {
  readonly type: typeof IN_MESSAGES.SEND_STATE_TO_RECONNECTED;
  playerID: string;
  isOwner: boolean;
  roomID: string;
  gameID?: string;
  started: boolean;
  gameVersion?: GameVersionKey;
  deck?: Card[];
  players: Record<string, Player>;
}

export interface ReconnectedToRoomMessage {
  readonly type: typeof IN_MESSAGES.RECONNECTED_TO_ROOM;
  roomID: string;
  playerID: string;
}

export interface CheckSetResultMessage {
  readonly type: typeof IN_MESSAGES.CHECK_SET_RESULT;
  isSet: boolean;
}

export interface ChangedGameStateMessage {
  readonly type: typeof IN_MESSAGES.CHANGED_GAME_STATE;
  gameID: string;
  deck: Card[];
  playerID: string;
  players: Record<string, Player>;
}

export interface GameOverMessage {
  readonly type: typeof IN_MESSAGES.GAME_OVER;
  gameID: string;
  deck: Card[];
  players: Record<string, Player>;
}

export interface ErrorMessage {
  readonly type: typeof IN_MESSAGES.ERROR;
  refType: keyof typeof OUT_MESSAGES;
  field: string;
  reason: string;
}

export type InMessage =
  (
    | StartedGameMessage
    | CreatedRoomMessage
    | JoinedRoomMessage
    | LeftRoomMessage
    | SendStateToReconnectedMessage
    | ReconnectedToRoomMessage
    | CheckSetResultMessage
    | ChangedGameStateMessage
    | GameOverMessage
    | ErrorMessage
  ) & {
    isProcessed?: boolean;
    error?: string | null;
  };

File: server/internal/messages/messages.go
package messages

import (
	"encoding/json"
	"server/internal/config"
	"server/internal/game"
	"server/internal/server"

	"github.com/google/uuid"
)

type InMessageType string
type InMessage struct {
	Type InMessageType `json:"type"`
}

const (
	CreateRoom InMessageType = "CREATE_ROOM"
	JoinRoom   InMessageType = "JOIN_ROOM"
	StartGame  InMessageType = "START_GAME"
	CheckSet   InMessageType = "CHECK_SET"
)

type StartGameMessage struct {
	InMessage
	GameVersion game.GameVersion `json:"gameVersion"`
	RoomID      uuid.UUID
}

type CreateRoomMessage struct {
	InMessage
	Nickname string `json:"nickname"`
}

type JoinRoomMessage struct {
	InMessage
	RoomID   uuid.UUID
	Nickname string `json:"nickname"`
}

type CheckSetMessage struct {
	InMessage
	CardIDs  []uuid.UUID `json:"cardIDs"`
	PlayerID uuid.UUID   `json:"playerID"`
	RoomID   uuid.UUID   `json:"roomID"`
	GameID   uuid.UUID   `json:"gameID"`
}

type Handler struct {
	ID       uuid.UUID
	Cfg      *config.Config
	Handlers map[InMessageType]InMessageHandler
}

func NewHandler(cfg *config.Config) *Handler {
	return &Handler{
		ID:       uuid.New(),
		Cfg:      cfg,
		Handlers: make(map[InMessageType]InMessageHandler),
	}
}

type InMessageHandler func(client *server.Client, rawMsg json.RawMessage) error

func (h *Handler) RegisterHandlers() {
	h.Handlers = map[InMessageType]InMessageHandler{
		CreateRoom: h.handleCreateRoom,
		JoinRoom:   h.handleJoinRoom,
		StartGame:  h.handleStartGame,
		CheckSet:   h.handleCheckSet,
	}
}

// Out

type OutMessageType string
type BaseOutMessage struct {
	Type OutMessageType `json:"type"`
}

const (
	CreatedRoom      OutMessageType = "CREATED_ROOM"
	JoinedRoom       OutMessageType = "JOINED_ROOM"
	StartedGame      OutMessageType = "STARTED_GAME"
	CheckSetResult   OutMessageType = "CHECK_SET_RESULT"
	ChangedGameState OutMessageType = "CHANGED_GAME_STATE"
	GameOver         OutMessageType = "GAME_OVER"
	ErrorOut         OutMessageType = "ERROR"
)

type CreatedRoomMessage struct {
	BaseOutMessage
	RoomID   uuid.UUID `json:"roomID"`
	PlayerID uuid.UUID `json:"playerID"`
}

type JoinedRoomMessage struct {
	BaseOutMessage
	RoomID   uuid.UUID `json:"roomID"`
	PlayerID uuid.UUID `json:"playerID"`
}

type StartedGameMessage struct {
	BaseOutMessage
	GameID      uuid.UUID                 `json:"gameID"`
	GameVersion game.GameVersion          `json:"gameVersion"`
	Deck        []game.Card               `json:"deck"`
	Players     map[uuid.UUID]game.Player `json:"players"`
}

type CheckSetResultMessage struct {
	BaseOutMessage
	IsSet bool `json:"isSet"`
}

type ChangedGameStateMessage struct {
	BaseOutMessage
	GameID  uuid.UUID                 `json:"gameID"`
	Deck    []game.Card               `json:"deck"`
	Players map[uuid.UUID]game.Player `json:"players"`
}

type GameOverMessage struct {
	BaseOutMessage
	GameID  uuid.UUID                 `json:"gameID"`
	Deck    []game.Card               `json:"deck"`
	Players map[uuid.UUID]game.Player `json:"players"`
}

// Error
type ErrorMessage struct {
	RefType InMessageType `json:"refType"`
	Field   string        `json:"field"`
	Reason  string        `json:"reason"`
}
